Add PackageManifest Class

AyanSinhaMahapatra opened the pull request:
Fixes #2747 

<!--
Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!
* [x] - Checked Box
* [ ] - Unchecked Box
-->

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

<!--
We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged**
if your tests fail at first!
If tests do fail, click on the red `X` to learn why by reading the logs.
Thanks!
-->

<!-- Don't forget to Signoff -->

AyanSinhaMahapatra committed:
Rename package attribute metafiles to file_patterns

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add PackageManifest class to models

Adds a mixin class PackageManifest to models which will have the base
implementations of Package Manifest specific methods for recognizing and
creating ecosystem specific package manifest objects, which will inherit
from this class and the ecosystem Package class.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Create PackageManifest classes for cocoapods

Creates PackageManifest classes for all individual types of cocoapods manifests
and overrides the methods for detection and PackageManifest creation.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Modify package classes to inherit PackageManifest mixin

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Create PackageManifest class fot aboutfiles

Creates PackacgeManifest class for aboutfiles and overrides
detection and PackageMamanifest creation methods.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Modify test expectations for recognize_package_manifests

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Create PackageManifest classes for NPM

Creates PackageManifest classes for NPM manifests
and overrides detection and PackageMamanifest creation methods.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Create PackageManifest classes for phpcomposer

Creates PackageManifest classes for all phpcomposer manifests
and overrides the methods for detection and PackageManifest creation.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Create PackageManifest class for Haxe

Creates PackageManifest classes for haxelib json manifests
and overrides the methods for detection and PackageManifest creation.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra commented on src/packagedcode/about.py:
```diff
@@ -47,53 +43,55 @@ def get_package_root(self, manifest_resource, codebase):
+        """
+        Return True if the file at ``location`` is likely a manifest of this type.
+        """
+        return (filetype.is_file(location) and location.lower().endswith(('.about',)))
```
No parentthesis here.

AyanSinhaMahapatra commented on tests/packagedcode/test_about.py:
```diff
@@ -17,14 +17,18 @@
 class TestAbout(PackageTester):
     test_data_dir = os.path.join(os.path.dirname(__file__), 'data')
 
+    def test_can_detect_aboutfile(self):
```
`aboutfile` -> `about_file`

AyanSinhaMahapatra committed:
Create PackageManifest classes for cargo

Creates PackageManifest classes for rust cargo manifests`
and overrides the methods for detection and PackageManifest creation

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add PackageManifest class for opam

Creates PackageManifest classes for opam manifests
and overrides the methods for detection and PackageManifest creation.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add PackageManifest class for bower

Creates PackageManifest class for bower json manifest
and overrides the methods for detection and PackageManifest creation.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Create PackageManifest class for freebsd

Creates PackageManifest class for freebsd compact manifest
and overrides the methods for detection and PackageManifest creation.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add PackageManifest classes for ruby manifests

Creates PackageManifest class for various ruby manifests
and overrides the methods for detection and PackageManifest creation.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add PackageManifest classes for pypi manifests

Creates PackageManifest class for various python manifests
and overrides the methods for detection and PackageManifest creation.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Create PackageManifest classes for golang manifests

Creates PackageManifest class for go.mod and go.sum manifests
and overrides the methods for detection and PackageManifest creation.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Create PackageManifest class for nuspec manifest

Creates PackageManifest class for nuspec manifest
and overrides the methods for detection and PackageManifest creation.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add PackageManifest classes for chef manifests

Creates PackageManifest classes for chef manifests
and overrides the methods for detection and PackageManifest creation.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Create PackageManifest classes for conda

Creates PackageManifest classes for conda manifests
and overrides the methods for detection and PackageManifest creation.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Create PackageManifest class for readme manifests

Creates PackageManifest class for readme manifests
and overrides the methods for detection and PackageManifest creation.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Create PackageManifest classes for pubspec

Creates PackageManifest classes for pubspec manifests
and overrides the methods for detection and PackageManifest creation.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Create PackageManifest classes for windows manifests

Creates PackageManifest classes for various windows manifests
and overrides the methods for detection and PackageManifest creation.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Modify PackageManifest classes in build

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Create PackageManifest class for rpm

Creates PackageManifest class for rpm manifests
and overrides the methods for detection and PackageManifest creation.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Modify PackageManifest class names and attributes

Modifies PackageManifest class names to be camelcase and deletes
attribute `manifest_type` from individual classes.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Remove duplicate models and update docstrings

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra commented:
@pombredanne This is all green and ready to review!

pombredanne commented on src/packagedcode/__init__.py:
```diff
@@ -38,73 +39,99 @@
+}
 # We cannot have two package classes with the same type
-if len(PACKAGES_BY_TYPE) != len(PACKAGE_TYPES):
+if len(PACKAGE_MANIFESTS_BY_TYPE) != len(PACKAGE_MANIFEST_TYPES):
```
In hindsight, I think that this should be in a function and lazily called and not on called on import; otherwise this slows down the startup of everything. It should be a validation function that should be called in a test rather than at every import. And may be this validation function could be called only in tests

AyanSinhaMahapatra replied:
This is now resolved in #2825

AyanSinhaMahapatra committed:
Remove recognize function from package class

Removes recognize function from Package Class, since
the recognize function is now at PackageManifest class
and this isn't used anywhere.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne reviewed (approved):
Merging at last! Thank you ++
